ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/participant/util/EventFilter.scala

```diff
@@ -26,9 +26,23 @@ object EventFilter {
     def isSubmitterSubscriber(submitterParty: Party): Boolean =
       transactionFilter.filtersByParty.contains(submitterParty)
 
-    private val subscribersByTemplateId: Map[Ref.Identifier, Set[Party]] = {
-      val (specificSubscriptions, globalSubscriptions) = getSpecificAndGlobalSubscriptions(
-        transactionFilter)
+    lazy val (specificSubscriptions, globalSubscriptions) = {
+      val specific = new ArrayBuffer[(Ref.Identifier, Party)]
+      val global = immutable.Set.newBuilder[Party]
+      transactionFilter.filtersByParty.foreach {
```

Comment:
Since this code is quite imperative in nature it may be worth considering an imperative idiom
```scala
for ((party, filters) <- transactionFilter) {
  if (filters.inclusive.isEmpty) global += party
  else for (tid <- inclusive.templateIds) {
    specific += tid -> party
  }
}
```

Reply:
Good idea. Addressed in 421b44e24.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2931/commits/421b44e244a3efabd7a0f4d90e9bc03ec6d291de